common/src/main/java/io/netty/util/internal/SpscLinkedQueue.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util.internal;
+
+import java.util.Collection;
+import java.util.Queue;
+
+/**
+ * A lock-free concurrent single-consumer single-producer {@link Queue}. This is
+ * a weakened version of MPSC, as we have no need for XCHG, an ordered store is enough.
+ *
+ * It allows only one producer thread to perform the following operations exclusively:
+ * <ul>
+ *     <li>{@link #offer(Object)}, {@link #add(Object)}, {@link #addAll(Collection)}</li>
+ *     <li>{@link #isEmpty()}</li>
+ * </ul>
+ * .. while only one consumer thread is allowed to perform the following operations exclusively:
+ * <ul>
+ *     <li>{@link #poll()} and {@link #remove()}</li>
+ *     <li>{@link #element()}, {@link #peek()}</li>
+ *     <li>{@link #remove(Object)}, {@link #removeAll(Collection)}, and {@link #retainAll(Collection)}</li>
+ *     <li>{@link #clear()}</li>
+ *     <li>{@link #iterator()}</li>
+ *     <li>{@link #toArray()} and {@link #toArray(Object[])}</li>
+ *     <li>{@link #contains(Object)} and {@link #containsAll(Collection)}</li>
+ *     <li>{@link #size()}</li>
+ * </ul>
+ *
+ * <strong>The behavior of this implementation is undefined if you perform these operations
+ * from multiple threads.</strong>
+ *
+ * @param <E>
+ */
+final class SpscLinkedQueue<E> extends MpscLinkedQueue<E> {
+
+    private static final long serialVersionUID = -8357404840053423795L;
+
+    @Override
+    protected void addNewTail(MpscLinkedQueueNode<E> newTail) {
+        MpscLinkedQueueNode<E> oldTail = tailRef();
```

Comment:
@normanmaurer - In order to do this the class hierarchy would need to be restructured (JCTools now has different base classes). This will be a bigger effort to update all of the JCTools based queuing stuff. Perhaps we could do it in phases ... just do Spsc (and required base classes) in this PR, then a followup PR to get in sync with JCTools?